Exception triggered on filter when setting composite agg after key

Requests using composite aggs with `afterKey` set to a missing field end up throwing an exception due to what looks like incorrect mapping.

And two entries inside "index-a" and no entries in "index-b":

```
POST index-a/_doc
{
  "bool" : true
}

POST index-a/_doc
{
  "bool" : false
}

POST index-b/_doc
{}
```
which results in the following mappings:

```
{
  "index-b" : {
    "mappings" : { }
  },
  "index-a" : {
    "mappings" : {
      "properties" : {
        "bool" : {
          "type" : "boolean"
        }
      }
    }
  }
}
```

Run a composite agg against it:
```
POST index-*/_search
{
  "size" : 0,
  "_source" : false,
  "aggregations" : {
    "groupby" : {
      "composite" : {
        "size" : 1,
        "sources" : [
          {
            "agg" : {
              "terms" : {
                "field" : "bool",
                "missing_bucket" : true,
                "order" : "asc"
              }
            }
          }
        ]
        ,"after": {"agg" : false}
      }
    }
  }
}
```

This works and should return something like:

```
{
  "took" : 0,
  "timed_out" : false,
  "_shards" : {
    "total" : 2,
    "successful" : 2,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 2,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "groupby" : {
      "after_key" : {
        "agg" : false
      },
      "buckets" : [
        {
          "key" : {
            "agg" : false
          },
          "doc_count" : 1
        }
      ]
    }
  }
}
```

Ask for the next composite agg page:
```
{
  "size" : 0,
  "_source" : false,
  "aggregations" : {
    "groupby" : {
      "composite" : {
        "size" : 1,
        "sources" : [
          {
            "agg" : {
              "terms" : {
                "field" : "bool",
                "missing_bucket" : true,
                "order" : "asc"
              }
            }
          }
        ],
        "after": {"agg" : false}
      }
    }
  }
}
```

this results in:
```
{
  "took" : 0,
  "timed_out" : false,
  "_shards" : {
    "total" : 2,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 1,
    "failures" : [
      {
        "shard" : 0,
        "index" : "index-b",
        "node" : "_ropmc3_StufvH-TmSMSPQ",
        "reason" : {
          "type" : "illegal_argument_exception",
          "reason" : "invalid value, expected string, got Boolean"
        }
      }
    ]
  },
  ...
```

The full underlying exception is:

```
org.elasticsearch.ElasticsearchException$1: invalid value, expected string, got Boolean
	at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:633)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.guessRootCauses(SearchPhaseExecutionException.java:150)
	at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:615)
	at org.elasticsearch.ElasticsearchException.generateFailureXContent(ElasticsearchException.java:561)
	at org.elasticsearch.rest.BytesRestResponse.build(BytesRestResponse.java:138)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:99)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:82)
	at org.elasticsearch.rest.action.RestActionListener.onFailure(RestActionListener.java:47)
	at org.elasticsearch.client.node.NodeClient.lambda$executeLocally$1(NodeClient.java:107)
	at org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:179)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:142)
	at org.elasticsearch.xpack.sql.execution.PlanExecutor.lambda$nextPage$2(PlanExecutor.java:113)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:142)
	at org.elasticsearch.xpack.sql.execution.search.CompositeAggCursor$1.onFailure(CompositeAggCursor.java:151)
	at org.elasticsearch.client.node.NodeClient.lambda$executeLocally$1(NodeClient.java:107)
	at org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:179)
	at org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:66)
	at org.elasticsearch.action.ActionListener$RunAfterActionListener.onFailure(ActionListener.java:322)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.raisePhaseFailure(AbstractSearchAsyncAction.java:660)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:636)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:372)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:669)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:440)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:289)
	at org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:66)
	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:48)
	at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:405)
	at org.elasticsearch.transport.TransportService$5.handleException(TransportService.java:625)
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1169)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1278)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1252)
	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:50)
	at org.elasticsearch.transport.TransportChannel.sendErrorResponse(TransportChannel.java:45)
	at org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:40)
	at org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:77)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:728)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)\r\nCaused by: java.lang.IllegalArgumentException: invalid value, expected string, got Boolean
	at org.elasticsearch.search.aggregations.bucket.composite.GlobalOrdinalValuesSource.setAfter(GlobalOrdinalValuesSource.java:98)
	at org.elasticsearch.search.aggregations.bucket.composite.CompositeValuesCollectorQueue.<init>(CompositeValuesCollectorQueue.java:77)
	at org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregator.<init>(CompositeAggregator.java:106)
	at org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationFactory.createInternal(CompositeAggregationFactory.java:37)
	at org.elasticsearch.search.aggregations.AggregatorFactory.create(AggregatorFactory.java:63)
	at org.elasticsearch.search.aggregations.AggregatorFactories.createSubAggregators(AggregatorFactories.java:187)
	at org.elasticsearch.search.aggregations.AggregatorFactories.createTopLevelAggregators(AggregatorFactories.java:197)
	at org.elasticsearch.search.aggregations.AggregationPhase.preProcess(AggregationPhase.java:39)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:138)
	at org.elasticsearch.indices.IndicesService.lambda$loadIntoContext$25(IndicesService.java:1409)
	at org.elasticsearch.indices.IndicesService.lambda$cacheShardLevelResult$26(IndicesService.java:1472)
	at org.elasticsearch.indices.IndicesRequestCache$Loader.load(IndicesRequestCache.java:164)
	at org.elasticsearch.indices.IndicesRequestCache$Loader.load(IndicesRequestCache.java:147)
	at org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:424)
	at org.elasticsearch.indices.IndicesRequestCache.getOrCompute(IndicesRequestCache.java:110)
	at org.elasticsearch.indices.IndicesService.cacheShardLevelResult(IndicesService.java:1478)
	at org.elasticsearch.indices.IndicesService.loadIntoContext(IndicesService.java:1403)
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:377)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:443)
	at org.elasticsearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:408)
	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:47)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	... 6 more"
```

This appears to have worked in 7.8 but occurs currently in master and 7.11.1

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java
@@ -92,7 +92,9 @@ class GlobalOrdinalValuesSource extends SingleDimensionValuesSource<BytesRef> {
         if (missingBucket && value == null) {
             afterValue = null;
             afterValueGlobalOrd = -1L;
-        } else if (value.getClass() == String.class) {
+        } else if (value.getClass() == String.class || (missingBucket && fieldType == null)) {
+            // the value might be not string if this field is missing in this shard but present in other shards
+            // and doesn't have a string type
             afterValue = format.parseBytesRef(value.toString());
         } else {
             throw new IllegalArgumentException("invalid value, expected string, got " + value.getClass().getSimpleName());

--- a/server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregatorTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregatorTests.java
@@ -516,7 +516,7 @@ public class CompositeAggregatorTests  extends AggregatorTestCase {
                     new TermsValuesSourceBuilder("long").field("long"),
                     new TermsValuesSourceBuilder("unmapped").field("unmapped").missingBucket(true)
                 )
-            ).aggregateAfter(Map.of("long", 1, "unmapped", randomFrom(randomBoolean(), 1, "b"))),
+            ).aggregateAfter(createAfterKey("long", 1, "unmapped", randomFrom(randomBoolean(), 1, "b"))),
             (result) -> {
                 assertEquals(2, result.getBuckets().size());
                 assertEquals("{long=4, unmapped=null}", result.afterKey().toString());

--- a/server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregatorTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregatorTests.java
@@ -432,6 +432,102 @@ public class CompositeAggregatorTests  extends AggregatorTestCase {
         );
     }
 
+    public void testUnmappedFieldWithLongs() throws Exception {
+        final List<Map<String, List<Object>>> dataset = new ArrayList<>();
+        dataset.addAll(
+            Arrays.asList(
+                createDocument("long", 1L),
+                createDocument("long", 3L),
+                createDocument("long", 1L),
+                createDocument("long", 4L),
+                createDocument("long", 3L)
+            )
+        );
+
+        testSearchCase(Arrays.asList(new MatchAllDocsQuery(), new DocValuesFieldExistsQuery("long")), dataset,
+            () -> new CompositeAggregationBuilder("name",
+                Arrays.asList(
+                    new TermsValuesSourceBuilder("unmapped").field("unmapped")
+                )
+            ),
+            (result) -> {
+                assertEquals(0, result.getBuckets().size());
+            }
+        );
+
+        testSearchCase(Arrays.asList(new MatchAllDocsQuery(), new DocValuesFieldExistsQuery("long")), dataset,
+            () -> new CompositeAggregationBuilder("name",
+                Arrays.asList(
+                    new TermsValuesSourceBuilder("unmapped").field("unmapped").missingBucket(true)
+                )
+            ),
+            (result) -> {
+                assertEquals(1, result.getBuckets().size());
+                assertEquals("{unmapped=null}", result.afterKey().toString());
+                assertEquals("{unmapped=null}", result.getBuckets().get(0).getKeyAsString());
+                assertEquals(5L, result.getBuckets().get(0).getDocCount());
+            }
+        );
+
+        testSearchCase(Arrays.asList(new MatchAllDocsQuery(), new DocValuesFieldExistsQuery("long")), dataset,
+            () -> new CompositeAggregationBuilder("name",
+                Arrays.asList(
+                    new TermsValuesSourceBuilder("unmapped").field("unmapped").missingBucket(true)
+                )).aggregateAfter(Collections.singletonMap("unmapped", null)),
+            (result) -> {
+                assertEquals(0, result.getBuckets().size());
+            }
+        );
+
+        testSearchCase(Arrays.asList(new MatchAllDocsQuery(), new DocValuesFieldExistsQuery("long")), dataset,
+            () -> new CompositeAggregationBuilder("name",
+                Arrays.asList(
+                    new TermsValuesSourceBuilder("long").field("long"),
+                    new TermsValuesSourceBuilder("unmapped").field("unmapped")
+                )
+            ),
+            (result) -> {
+                assertEquals(0, result.getBuckets().size());
+            }
+        );
+
+        testSearchCase(Arrays.asList(new MatchAllDocsQuery(), new DocValuesFieldExistsQuery("long")), dataset,
+            () -> new CompositeAggregationBuilder("name",
+                Arrays.asList(
+                    new TermsValuesSourceBuilder("long").field("long"),
+                    new TermsValuesSourceBuilder("unmapped").field("unmapped").missingBucket(true)
+                )
+            ),
+            (result) -> {
+                assertEquals(3, result.getBuckets().size());
+                assertEquals("{long=4, unmapped=null}", result.afterKey().toString());
+                assertEquals("{long=1, unmapped=null}", result.getBuckets().get(0).getKeyAsString());
+                assertEquals(2L, result.getBuckets().get(0).getDocCount());
+                assertEquals("{long=3, unmapped=null}", result.getBuckets().get(1).getKeyAsString());
+                assertEquals(2L, result.getBuckets().get(1).getDocCount());
+                assertEquals("{long=4, unmapped=null}", result.getBuckets().get(2).getKeyAsString());
+                assertEquals(1L, result.getBuckets().get(2).getDocCount());
+            }
+        );
+
+        testSearchCase(Arrays.asList(new MatchAllDocsQuery(), new DocValuesFieldExistsQuery("long")), dataset,
+            () -> new CompositeAggregationBuilder("name",
+                Arrays.asList(
+                    new TermsValuesSourceBuilder("long").field("long"),
+                    new TermsValuesSourceBuilder("unmapped").field("unmapped").missingBucket(true)
+                )
+            ).aggregateAfter(Map.of("long", 1, "unmapped", randomFrom(randomBoolean(), 1, "b"))),
+            (result) -> {
+                assertEquals(2, result.getBuckets().size());
+                assertEquals("{long=4, unmapped=null}", result.afterKey().toString());
+                assertEquals("{long=3, unmapped=null}", result.getBuckets().get(0).getKeyAsString());
+                assertEquals(2L, result.getBuckets().get(0).getDocCount());
+                assertEquals("{long=4, unmapped=null}", result.getBuckets().get(1).getKeyAsString());
+                assertEquals(1L, result.getBuckets().get(1).getDocCount());
+            }
+        );
+    }
+
     public void testWithKeyword() throws Exception {
         final List<Map<String, List<Object>>> dataset = new ArrayList<>();
         dataset.addAll(

